File: sdks/official/fraiseql-fsharp/src/FraiseQL/MutationBuilder.fs
namespace FraiseQL

/// Pipe-friendly builder for constructing <see cref="MutationDefinition"/> values.
///
/// Example:
/// <code>
/// MutationBuilder.mutation "createAuthor"
/// |> MutationBuilder.returnType "Author"
/// |> MutationBuilder.sqlSource "fn_create_author"
/// |> MutationBuilder.operation "insert"
/// |> MutationBuilder.register
/// </code>
module MutationBuilder =

    /// Internal accumulator state for building a <see cref="MutationDefinition"/>.
    type MutationState =
        {
            name: string
            returnType: string
            sqlSource: string
            operation: string
            arguments: ArgumentDefinition list
            description: string option
            rest: RestConfig option
            injectParams: Map<string, string>
            requiresRole: string option
            invalidatesViews: string list
            invalidatesFactTables: string list
        }

    /// Creates a new <see cref="MutationState"/> for the given mutation name.
    let mutation (name: string) : MutationState =
        {
            name = name
            returnType = ""
            sqlSource = ""
            operation = "custom"
            arguments = []
            description = None
            rest = None
            injectParams = Map.empty
            requiresRole = None
            invalidatesViews = []
            invalidatesFactTables = []
        }

    /// Sets the GraphQL return type for this mutation.
    let returnType (t: string) (s: MutationState) : MutationState = { s with returnType = t }

    /// Sets the SQL function backing this mutation.
    let sqlSource (src: string) (s: MutationState) : MutationState = { s with sqlSource = src }

    /// Sets the operation kind: "insert", "update", "delete", or "custom".
    let operation (op: string) (s: MutationState) : MutationState = { s with operation = op }

    /// Sets the optional human-readable description.
    let description (d: string) (s: MutationState) : MutationState =
        { s with description = Some d }

    /// Adds an argument to this mutation.
    let withArgument (name: string) (type_: string) (isNullable: bool) (s: MutationState) : MutationState =
        let arg: ArgumentDefinition = { name = name; type_ = type_; nullable = isNullable }
        { s with arguments = s.arguments @ [ arg ] }

    /// Sets the optional REST endpoint annotation.
    let rest (cfg: RestConfig) (s: MutationState) : MutationState = { s with rest = Some cfg }

    /// Declares a server-injected parameter, not exposed as a GraphQL argument.
    /// `source` is of the form `"jwt:&lt;claim&gt;"`.
    let inject (parameter: string) (source: string) (s: MutationState) : MutationState =
        { s with injectParams = Map.add parameter source s.injectParams }

    /// Restricts this mutation to callers holding the given role.
    let requiresRole (role: string) (s: MutationState) : MutationState =
        { s with requiresRole = Some role }

    /// Declares views whose cached query results must be invalidated after this mutation.
    let invalidatesViews (views: string list) (s: MutationState) : MutationState =
        { s with invalidatesViews = s.invalidatesViews @ views }

    /// Declares fact tables whose cached aggregates must be invalidated after this mutation.
    let invalidatesFactTables (tables: string list) (s: MutationState) : MutationState =
        { s with invalidatesFactTables = s.invalidatesFactTables @ tables }

    /// Converts the accumulated state into a <see cref="MutationDefinition"/>.
    /// Raises <see cref="System.InvalidOperationException"/> when required fields are missing.
    let toDefinition (s: MutationState) : MutationDefinition =
        if s.returnType = "" then
            raise (
                System.InvalidOperationException(sprintf "Mutation '%s' has no returnType" s.name)
            )

        if s.sqlSource = "" then
            raise (
                System.InvalidOperationException(sprintf "Mutation '%s' has no sqlSource" s.name)
            )

        {
            name = s.name
            return_type = s.returnType
            sql_source = s.sqlSource
            operation = s.operation
            arguments = s.arguments
            description = s.description
            rest = s.rest
            cascade = None
            inject_params = (if Map.isEmpty s.injectParams then None else Some s.injectParams)
            requires_role = s.requiresRole
            invalidates_views =
                (if List.isEmpty s.invalidatesViews then None else Some s.invalidatesViews)
            invalidates_fact_tables =
                (if List.isEmpty s.invalidatesFactTables then None else Some s.invalidatesFactTables)
        }

    /// Converts the state to a <see cref="MutationDefinition"/> and registers it in <see cref="SchemaRegistry"/>.
    let register (s: MutationState) : unit = SchemaRegistry.registerMutation (toDefinition s)

File: sdks/official/fraiseql-csharp/src/FraiseQL/Builders/MutationBuilder.cs
using FraiseQL.Models;
using FraiseQL.Registry;

namespace FraiseQL.Builders;

/// <summary>
/// Fluent builder for constructing a <see cref="IntermediateMutation"/> and optionally
/// registering it in the <see cref="SchemaRegistry"/>.
/// </summary>
/// <example>
/// <code>
/// MutationBuilder.Mutation("createAuthor")
///     .ReturnType("Author")
///     .SqlSource("fn_create_author")
///     .Operation("insert")
///     .Argument("name", "String")
///     .Register();
/// </code>
/// </example>
public sealed class MutationBuilder
{
    private static readonly HashSet<string> ValidOperations =
        new(StringComparer.OrdinalIgnoreCase) { "insert", "update", "delete", "upsert" };

    private readonly string _name;
    private string _returnType = string.Empty;
    private string _sqlSource = string.Empty;
    private string _operation = string.Empty;
    private string? _description;
    private string? _restPath;
    private string? _restMethod;
    private readonly List<IntermediateArgument> _arguments = new();

    private MutationBuilder(string name) => _name = name;

    /// <summary>Creates a new <see cref="MutationBuilder"/> for a mutation with the given name.</summary>
    /// <param name="name">The GraphQL mutation name.</param>
    /// <returns>A new builder instance.</returns>
    public static MutationBuilder Mutation(string name) => new(name);

    /// <summary>Sets the GraphQL return type name.</summary>
    /// <param name="type">The return type name (e.g. <c>"Author"</c>).</param>
    /// <returns>This builder for chaining.</returns>
    public MutationBuilder ReturnType(string type) { _returnType = type; return this; }

    /// <summary>Sets the backing SQL function name.</summary>
    /// <param name="source">The SQL function name (e.g. <c>"fn_create_author"</c>).</param>
    /// <returns>This builder for chaining.</returns>
    public MutationBuilder SqlSource(string source) { _sqlSource = source; return this; }

    /// <summary>
    /// Sets the mutation operation kind.
    /// </summary>
    /// <param name="operation">One of <c>"insert"</c>, <c>"update"</c>, <c>"delete"</c>, or <c>"upsert"</c>.</param>
    /// <returns>This builder for chaining.</returns>
    /// <exception cref="ArgumentException">Thrown when <paramref name="operation"/> is not a valid operation.</exception>
    public MutationBuilder Operation(string operation)
    {
        if (!ValidOperations.Contains(operation))
            throw new ArgumentException(
                $"Invalid operation '{operation}'. Must be one of: insert, update, delete, upsert.",
                nameof(operation));
        _operation = operation;
        return this;
    }

    /// <summary>Sets a human-readable description for this mutation.</summary>
    /// <param name="desc">The description text.</param>
    /// <returns>This builder for chaining.</returns>
    public MutationBuilder Description(string desc) { _description = desc; return this; }

    /// <summary>Sets the REST endpoint path for this mutation.</summary>
    /// <param name="path">The REST path (e.g. <c>"/api/users"</c>).</param>
    /// <returns>This builder for chaining.</returns>
    public MutationBuilder RestPath(string path) { _restPath = path; return this; }

    /// <summary>Sets the HTTP method for the REST endpoint. Defaults to POST for mutations.</summary>
    /// <param name="method">The HTTP method (GET, POST, PUT, PATCH, DELETE).</param>
    /// <returns>This builder for chaining.</returns>
    public MutationBuilder RestMethod(string method) { _restMethod = method; return this; }

    /// <summary>Adds a typed argument to this mutation.</summary>
    /// <param name="name">The argument name.</param>
    /// <param name="type">The GraphQL type name.</param>
    /// <param name="nullable">Whether the argument accepts <c>null</c>.</param>
    /// <returns>This builder for chaining.</returns>
    public MutationBuilder Argument(string name, string type, bool nullable = false)
    {
        _arguments.Add(new IntermediateArgument(name, type, nullable));
        return this;
    }

    /// <summary>
    /// Builds the <see cref="IntermediateMutation"/> from the current configuration.
    /// </summary>
    /// <returns>The constructed mutation.</returns>
    /// <exception cref="InvalidOperationException">
    /// Thrown when <c>ReturnType</c>, <c>SqlSource</c>, or <c>Operation</c> has not been set.
    /// </exception>
    public IntermediateMutation Build()
    {
        if (string.IsNullOrEmpty(_returnType))
            throw new InvalidOperationException(
                $"MutationBuilder: ReturnType must be set before Build() (mutation: '{_name}')");
        if (string.IsNullOrEmpty(_sqlSource))
            throw new InvalidOperationException(
                $"MutationBuilder: SqlSource must be set before Build() (mutation: '{_name}')");
        if (string.IsNullOrEmpty(_operation))
            throw new InvalidOperationException(
                $"MutationBuilder: Operation must be set before Build() (mutation: '{_name}')");

        RestAnnotation? rest = _restPath != null
            ? new RestAnnotation(_restPath, (_restMethod ?? "POST").ToUpperInvariant())
            : null;

        return new IntermediateMutation(
            Name: _name,
            ReturnType: _returnType,
            SqlSource: _sqlSource,
            Operation: _operation,
            Arguments: _arguments.AsReadOnly(),
            Description: _description,
            Rest: rest);
    }

    /// <summary>
    /// Builds the mutation and registers it in <see cref="SchemaRegistry.Instance"/>.
    /// Equivalent to calling <c>SchemaRegistry.Instance.RegisterMutation(Build())</c>.
    /// </summary>
    public void Register() => SchemaRegistry.Instance.RegisterMutation(Build());

    /// <summary>Returns the mutation name set on this builder.</summary>
    /// <returns>The mutation name.</returns>
    internal string GetName() => _name;
}
